### ISSUE TITLE
Improve sporadically failing integration tests to surface GSI timeout related issues

### ISSUE THREAD

**Issue by adamcfraser** [couchbase/sync_gateway#3570]
It's expected that most of the sporadically failing integration tests are due to GSI timeouts associated with post-flush index recovery.  However, most of those tests don't do a good job of exposing the fact that the underlying issue was a GSI issue.

Improve the tests (listed in #3539) to provide clarity on underlying root cause.

Related to https://github.com/couchbase/sync_gateway/pull/3544/files

**Comment by tleyden** [comment]
Current sporadic failures + logs:

## TestAccessFunctionDb

```
=== RUN   TestAccessFunctionDb
2018-05-14 15:07:57.086096 I | Flushing bucket test_data_bucket
2018-05-14T15:08:06.264-07:00 [INF] GoCBCustomSGTranscoder Opening Couchbase database test_data_bucket on <http://integration-cbs-55.sc.couchbase.com:8091> as user "test_data_bucket"
2018-05-14T15:08:06.594-07:00 [INF] Design docs successfully created for view version 2.1.
2018-05-14T15:08:06.594-07:00 [INF] Initializing indexes with numReplicas: 0...
2018-05-14T15:08:08.168-07:00 [INF] Verifying index availability for bucket test_data_bucket...
2018-05-14T15:08:08.182-07:00 [INF] Indexes ready for bucket test_data_bucket.
2018-05-14T15:08:08.183-07:00 [INF] Initializing changes cache for database db with sequence: 0
2018-05-14T15:08:08.263-07:00 [INF] Using metadata purge interval of 3.00 days for tombstone compaction.
_time=2018-05-14T15:08:08.376-07:00 _level=INFO _msg=Using plain authentication for user test_data_bucket 

	database_test.go:1120
	assert.DeepEquals(t, user.Channels(), expected)	

	expected: channels.TimedSet{"Hulu":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x1}, "Netflix":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x1}, "!":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x1}}
	     got: channels.TimedSet{"!":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x1}, "Netflix":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x1}}

	database_test.go:1123
	assert.DeepEquals(t, user.InheritedChannels(), expected)	

	expected: channels.TimedSet{"Hulu":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x1}, "Netflix":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x1}, "!":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x1}, "CrunchyRoll":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x2}}
	     got: channels.TimedSet{"Netflix":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x1}, "!":channels.VbSequence{VbNo:(*uint16)(nil), Sequence:0x1}}
2018-05-14T15:08:08.414-07:00 [WRN] Error processing DCP stream - will attempt to restart/reconnect if appropriate: pkt.Receive, err: read tcp 172.23.105.143:34866->172.23.106.124:11210: use of closed network connection. -- base.(*DCPReceiver).OnError() at dcp_feed.go:137
2018-05-14T15:08:08.414-07:00 [WRN] Error closing GoCB bucket: connection shut down. -- base.(*CouchbaseBucketGoCB).Close() at bucket_gocb.go:2273
--- FAIL: TestAccessFunctionDb (11.34s)
	assert.go:31: 
	assert.go:31: 
FAIL
exit status 1
```

## TestAllDocsChannelsAfterChannelMove

Repro with:

```
SG_TEST_BACKING_STORE=Couchbase SG_TEST_USE_XATTRS=true repeat 10 ./test.sh -v -run TestAllDocsChannelsAfterChannelMove$
```

Logs:

```
=== RUN   TestAllDocsChannelsAfterChannelMove
2018-05-14 15:07:08.741257 I | Flushing bucket test_data_bucket
2018-05-14T15:07:15.456-07:00 [INF] GoCBCustomSGTranscoder Opening Couchbase database test_data_bucket on <http://couchbase-server:8091> as user "test_data_bucket"
2018-05-14T15:07:15.516-07:00 [INF] Opening db /db as bucket "test_data_bucket", pool "default", server <http://couchbase-server:8091>
2018-05-14T15:07:15.516-07:00 [INF] GoCBCustomSGTranscoder Opening Couchbase database test_data_bucket on <http://couchbase-server:8091> as user "test_data_bucket"
2018-05-14T15:07:15.538-07:00 [INF] Initializing indexes with numReplicas: 0...

2018-05-14T15:07:16.185-07:00 [INF] Verifying index availability for bucket test_data_bucket...
2018-05-14T15:07:16.196-07:00 [INF] Indexes ready for bucket test_data_bucket.
2018-05-14T15:07:16.199-07:00 [INF] Initializing changes cache for database db with sequence: 0
2018-05-14 15:07:16.242489 I | bucketDataSource.Start() called
2018-05-14T15:07:16.251-07:00 [INF] Using metadata purge interval of 3.00 days for tombstone compaction.
2018-05-14T15:07:16.269-07:00 [WRN] QueryAccess called with empty username - returning empty result iterator -- db.(*DatabaseContext).QueryAccess() at query.go:239
2018-05-14T15:07:16.269-07:00 [WRN] QueryRoleAccess called with empty username -- db.(*DatabaseContext).QueryRoleAccess() at query.go:260
_time=2018-05-14T15:07:16.288-07:00 _level=INFO _msg=Using plain authentication for user test_data_bucket
2018-05-14T15:07:21.782-07:00 [WRN] Error when querying index using statement: [SELECT META(`test_data_bucket`).id as id, meta(`test_data_bucket`).xattrs._sync.rev as r, meta(`test_data_bucket`).xattrs._sync.sequence as s, meta(`test_data_bucket`).xattrs._sync.channels as c FROM `test_data_bucket` WHERE meta(`test_data_bucket`).xattrs._sync.sequence > 0 AND META(`test_data_bucket`).id NOT LIKE '\\_sync:%' AND meta(`test_data_bucket`).xattrs._sync IS NOT MISSING AND (meta(`test_data_bucket`).xattrs._sync.flags IS MISSING OR BITTEST(meta(`test_data_bucket`).xattrs._sync.flags,1) = false) ORDER BY META(`test_data_bucket`).id] -- base.(*CouchbaseBucketGoCB).Query() at bucket_n1ql.go:63
2018-05-14T15:07:21.782-07:00 [INF] HTTP: #002:     --> 500 Internal error: [5001] Panic: runtime error: index out of range  (5492.2 ms)
2018-05-14 15:07:21.782802 I | Admin response = {"rows":[
{"error":"Internal Server Error","reason":"Internal error: [5001] Panic: runtime error: index out of range"}

	api_test.go:2699
	assert.Equals(t, err, nil)

	expected: <nil>
	     got: &json.SyntaxError{msg:"unexpected end of JSON input", Offset:118}

	api_test.go:2700
	assert.Equals(t, len(allDocsResult.Rows), 1)

	expected: 1
	     got: 0
--- FAIL: TestAllDocsChannelsAfterChannelMove (13.06s)
	assert.go:31:
	assert.go:31:
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 19 [running]:
testing.tRunner.func1(0xc4202220f0)
	/usr/local/go/src/testing/testing.go:711 +0x2d2
panic(0x1884920, 0x2027d80)
	/usr/local/go/src/runtime/panic.go:491 +0x283
github.com/couchbase/sync_gateway/rest.TestAllDocsChannelsAfterChannelMove(0xc4202220f0)
	/Users/tleyden/Development/sync_gateway/godeps/src/github.com/couchbase/sync_gateway/rest/api_test.go:2701 +0x1454
testing.tRunner(0xc4202220f0, 0x19be260)
	/usr/local/go/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:789 +0x2de
exit status 2
FAIL	github.com/couchbase/sync_gateway/rest	13.084s
```

**Comment by tleyden** [comment]
In a recent run http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/691/consoleFull

All errors were due to "view_undefined" errors:

```
2018-05-18T18:42:14.648-07:00 [INF] HTTP: #3991:     --> 500 Internal error: unknown_error - view_undefined  (44.0 ms)
--- FAIL: TestViewQueryMultipleViews (3.70s)
	changes_api_test.go:2176: Unexpected error: Got response code: 500 from view call.  Expected 200.
=== RUN   TestViewQueryUserAccess
```

```
2018-05-18T18:43:55.924-07:00 [INF] HTTP: #4021:     --> 500 Internal error: unknown_error - view_undefined  (14.7 ms)
--- FAIL: TestAdminReduceViewQuery (29.92s)
	changes_api_test.go:2176: Unexpected error: Got response code: 500 from view call.  Expected 200.
```

```
2018-05-18T18:47:14.307-07:00 [INF] HTTP: #4086:     --> 500 Internal error: unknown_error - view_undefined  (23.0 ms)
--- FAIL: TestAdminGroupLevelReduceSumQuery (31.83s)
	changes_api_test.go:2176: Unexpected error: Got response code: 500 from view call.  Expected 200.
```

**Comment by tleyden** [comment]
Looks like it's related to https://github.com/couchbase/sync_gateway/issues/3346.  Will look into why this fix isn't covering these cases.

**Comment by tleyden** [comment]
In the `TestViewQueryMultipleViews` test, it looks like `InitializeIndexes()` (which waits for views to be ready) is not being called, since `DatabaseConfig.UseViews = false`

Trying this workaround:

```
func TestViewQueryMultipleViews(t *testing.T) {
	var rt RestTester
	rt.DatabaseConfig.UseViews = true  // <-- added this
	defer rt.Close()
```

**Comment by tleyden** [comment]
After adding a fix for the view_undefined errors in https://github.com/couchbase/sync_gateway/commit/9330435744f800a2664c776f423aa425a5fea88a, still seeing failures in Jenkins integration tests: http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/693/consoleFull


```
2018-05-20T08:35:22.732-07:00 [INF] WaitForIndexEmpty found non-zero results (1).  Retrying until the GSI index is empty.
2018-05-20T08:35:24.739-07:00 [INF] WaitForIndexEmpty found non-zero results (1).  Retrying until the GSI index is empty.
2018-05-20T08:35:26.749-07:00 [INF] WaitForIndexEmpty found non-zero results (1).  Retrying until the GSI index is empty.
2018-05-20T08:35:26.749-07:00 [WRN] RetryLoop for Wait for index to be empty giving up after 31 attempts -- base.RetryLoop() at util.go:344
2018-05-20 08:35:26.749836 I | Error waiting for GSI indexes to rollback:<nil>
FAIL	github.com/couchbase/sync_gateway/db	391.637s
Build was aborted
```

The GSI index wasn't empty after approximately 1 minute of waiting.  I tripled the timeout in 901055d1ffe25ddb917bf1c9d5a150f82ca961fe and kicked off re-runs.

**Comment by tleyden** [comment]
Full jenkins integration test completed OK: 

* xattrs=false: http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/694/
* xattrs=true: http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/695/

**Comment by tleyden** [comment]
Another failure in http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/696/console:

```
2018-05-20T11:56:58.399-07:00 [INF] WaitForIndexEmpty found non-zero results (1).  Retrying until the GSI index is empty.
2018-05-20T11:57:00.408-07:00 [INF] WaitForIndexEmpty found non-zero results (1).  Retrying until the GSI index is empty.
2018-05-20T11:57:02.416-07:00 [INF] WaitForIndexEmpty found non-zero results (1).  Retrying until the GSI index is empty.
2018-05-20T11:57:02.416-07:00 [WRN] RetryLoop for Wait for index to be empty giving up after 101 attempts -- base.RetryLoop() at util.go:344
2018-05-20 11:57:02.416472 I | Error waiting for GSI indexes to rollback: RetryLoop for Wait for index to be empty giving up after 101 attempts
FAIL	github.com/couchbase/sync_gateway/db	1495.265s
```

I think the best solution might be to:

1. Try to wait for the index to be empty
1. If not empty after 60 seconds, drop the index and retry

**Comment by tleyden** [comment]
> Try to wait for the index to be empty
> If not empty after 60 seconds, drop the index and retry

Implemented in 41c84b3026ccfe1a88f4d615a450b68abdae7abb

**Comment by tleyden** [comment]
Now seeing errors:

```
--- FAIL: TestSyncFnOnPush (67.93s)
panic: Error checking if bucket exists.  Spec: {Server:http://integration-cbs-55.sc.couchbase.com:8091 PoolName: BucketName:test_data_bucket FeedType: Auth:{Username:test_data_bucket Password:password BucketName:test_data_bucket} CouchbaseDriver:GoCBCustomSGTranscoder Certpath: Keypath: CACertPath: KvTLSPort:0 MaxNumRetries:0 InitialRetrySleepTimeMS:0 UseXattrs:true ViewQueryTimeoutSecs:<nil> BucketOpTimeout:<nil>} err: There are already 1 open buckets with name: test_data_bucket.  The tests expect all buckets to be closed. [recovered]
	panic: Error checking if bucket exists.  Spec: {Server:http://integration-cbs-55.sc.couchbase.com:8091 PoolName: BucketName:test_data_bucket FeedType: Auth:{Username:test_data_bucket Password:password BucketName:test_data_bucket} CouchbaseDriver:GoCBCustomSGTranscoder Certpath: Keypath: CACertPath: KvTLSPort:0 MaxNumRetries:0 InitialRetrySleepTimeMS:0 UseXattrs:true ViewQueryTimeoutSecs:<nil> BucketOpTimeout:<nil>} err: There are already 1 open buckets with name: test_data_bucket.  The tests expect all buckets to be closed.
```

(in http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/704/consoleFull)

**Comment by tleyden** [comment]
Ok, digging in some more, it's due to the fact that the GSI index didn't empty in time, so it dropped the index and retried, but ended up trying to open a second bucket connection without closing the first:

```
018-05-20T17:32:52.031-07:00 [INF] WaitForIndexEmpty found non-zero results (3).  Retrying until the GSI index is empty.
2018-05-20T17:32:54.039-07:00 [INF] WaitForIndexEmpty found non-zero results (3).  Retrying until the GSI index is empty.
2018-05-20T17:32:56.048-07:00 [INF] WaitForIndexEmpty found non-zero results (3).  Retrying until the GSI index is empty.
2018-05-20T17:32:58.056-07:00 [INF] WaitForIndexEmpty found non-zero results (3).  Retrying until the GSI index is empty.
2018-05-20T17:33:00.063-07:00 [INF] WaitForIndexEmpty found non-zero results (3).  Retrying until the GSI index is empty.
2018-05-20T17:33:00.063-07:00 [WRN] RetryLoop for Wait for index to be empty giving up after 31 attempts -- base.RetryLoop() at util.go:344
2018-05-20T17:33:00.063-07:00 [INF] Error WaitForIndexEmpty: RetryLoop for Wait for index to be empty giving up after 31 attempts.  Drop indexes and retry
2018-05-20 17:33:00.335539 I | Dropping index sg_allDocs_x1...
2018-05-20 17:33:00.335750 I | Dropping index sg_access_x1...
2018-05-20 17:33:00.335837 I | Dropping index sg_channels_x1...
2018-05-20 17:33:00.335926 I | Dropping index sg_tombstones_x1...
2018-05-20 17:33:00.336195 I | Dropping index sg_roleAccess_x1...
2018-05-20 17:33:00.336277 I | Dropping index sg_syncDocs_x1...
2018-05-20 17:33:01.459480 I | ...successfully dropped index sg_access_x1
2018-05-20 17:33:02.530000 I | ...successfully dropped index sg_channels_x1
2018-05-20 17:33:03.619371 I | ...successfully dropped index sg_allDocs_x1
2018-05-20 17:33:04.683705 I | ...successfully dropped index sg_roleAccess_x1
2018-05-20 17:33:05.733934 I | ...successfully dropped index sg_syncDocs_x1
2018-05-20 17:33:06.747071 I | ...successfully dropped index sg_tombstones_x1
--- FAIL: TestSyncFnOnPush (67.93s)
panic: Error checking if bucket exists.  Spec: {Server:http://integration-cbs-55.sc.couchbase.com:8091 PoolName: BucketName:test_data_bucket FeedType: Auth:{Username:test_data_bucket Password:password BucketName:test_data_bucket} CouchbaseDriver:GoCBCustomSGTranscoder Certpath: Keypath: CACertPath: KvTLSPort:0 MaxNumRetries:0 InitialRetrySleepTimeMS:0 UseXattrs:true ViewQueryTimeoutSecs:<nil> BucketOpTimeout:<nil>} err: There are already 1 open buckets with name: test_data_bucket.  The tests expect all buckets to be closed. [recovered]
	panic: Error checking if bucket exists.  Spec: {Server:http://integration-cbs-55.sc.couchbase.com:8091 PoolName: BucketName:test_data_bucket FeedType: Auth:{Username:test_data_bucket Password:password BucketName:test_data_bucket} CouchbaseDriver:GoCBCustomSGTranscoder Certpath: Keypath: CACertPath: KvTLSPort:0 MaxNumRetries:0 InitialRetrySleepTimeMS:0 UseXattrs:true ViewQueryTimeoutSecs:<nil> BucketOpTimeout:<nil>} err: There are already 1 open buckets with name: test_data_bucket.  The tests expect all buckets to be closed.
```

**Comment by tleyden** [comment]
Fix attempt for issue mentioned in previous comment: 2723eceb292aab9bf3ab3f5313af7f6e39a97d5a

**Comment by tleyden** [comment]
xattrs=true full suite passed: http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/705/parameters/
xattrs=false full suite in progress: http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/706/parameters/

**Comment by tleyden** [comment]
xattrs=false http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/706/parameters/ had a failure:

```
--- FAIL: TestChangesLoopingWhenLowSequence (69.39s)
panic: Error from AddDatabaseFromConfig: Unable to install index syncDocs: Error installing Couchbase index: sg_syncDocs_1: Unable to create indexes with the specified number of replicas (1).  Increase the number of index nodes, or modify 'num_index_replicas' in your Sync Gateway database config. [recovered]
	panic: Error from AddDatabaseFromConfig: Unable to install index syncDocs: Error installing Couchbase index: sg_syncDocs_1: Unable to create indexes with the specified number of replicas (1).  Increase the number of index nodes, or modify 'num_index_replicas' in your Sync Gateway database config.

goroutine 24369 [running]:
testing.tRunner.func1(0xc425e30340)
	/root/.gvm/gos/go1.8.5/src/testing/testing.go:622 +0x29d
panic(0xca8960, 0xc42665c090)
```

**Comment by tleyden** [comment]
Fixed in 6b4c8e5415a2758b83aedc6827f979dccd2d43af
